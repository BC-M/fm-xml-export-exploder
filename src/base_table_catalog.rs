use quick_xml::events::{BytesStart, Event};
use quick_xml::reader::Reader;
use std::collections::HashMap;
use std::io::{BufRead, Read};

use crate::utils::attributes::get_attributes;

pub fn parse_base_table_catalog<R: Read + BufRead>(
    reader: &mut Reader<R>,
    _: &BytesStart,
) -> HashMap<String, String> {
    let mut depth = 1;
    let mut table_name_id_map: HashMap<String, String> = HashMap::new();

    let mut buf = Vec::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Err(e) => {
                println!("Error {}", e);
                break;
            }
            Ok(Event::Eof) => break,
            Ok(Event::Start(e)) => {
                depth += 1;

                if depth == 2 && e.name().as_ref() == b"BaseTable" {
                    let mut id = String::new();
                    let mut name = String::new();

                    for attr in get_attributes(&e).unwrap() {
                        match attr.0.as_str() {
                            "id" => id = attr.1.to_string(),
                            "name" => name = attr.1.to_string(),
                            _ => {}
                        }
                    }

                    table_name_id_map.insert(name, id);
                }
            }
            Ok(Event::End(_)) => {
                depth -= 1;
                if depth == 0 {
                    break;
                }
            }
            _ => {}
        }

        buf.clear()
    }

    table_name_id_map
}
