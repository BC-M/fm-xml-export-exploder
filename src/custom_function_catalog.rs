use crate::{escape_filename, join_scope_id_and_name};
use quick_xml::events::{BytesStart, Event};
use quick_xml::reader::Reader;
use std::io::{BufRead, Read};
use std::path::Path;

use crate::utils::attributes::get_attributes;
use crate::utils::xml_utils::{cdata_to_string, local_name_to_string};
use crate::utils::{initialize_out_dir, write_text_file};

#[derive(Debug, Default)]
struct CustomFunctionInfo {
    id: String,
    name: String,
    content: String,
}

pub fn xml_explode_custom_function_catalog<R: Read + BufRead>(
    reader: &mut Reader<R>,
    _: &BytesStart,
    out_dir_path: &Path,
    fm_file_name: &str,
) {
    let out_dir_path = out_dir_path.join("custom_functions").join(fm_file_name);
    initialize_out_dir(&out_dir_path);

    let mut custom_function_info = CustomFunctionInfo::default();

    let mut depth = 1;
    let mut buf = Vec::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Err(_) => {
                break;
            }
            Ok(Event::Eof) => break,
            Ok(Event::Start(e)) => {
                depth += 1;
                if depth < 3 {
                    continue;
                } else if depth == 3 {
                    custom_function_info.id.clear();
                    custom_function_info.name.clear();
                    custom_function_info.content.clear();
                } else if depth == 4 && e.name().as_ref() == b"CustomFunctionReference" {
                    for attr in get_attributes(&e).unwrap() {
                        match attr.0.as_str() {
                            "id" => custom_function_info.id = attr.1.to_string(),
                            "name" => custom_function_info.name = attr.1.to_string(),
                            _ => {}
                        }
                    }
                }
            }
            Ok(Event::End(e)) => {
                depth -= 1;
                if depth == 0 {
                    break;
                } else if depth < 2 {
                    continue;
                } else if depth == 2
                    && local_name_to_string(e.name().as_ref()) == "CustomFunctionCalc"
                {
                    write_custom_function_to_file(&out_dir_path, &custom_function_info);
                    custom_function_info.id.clear();
                    custom_function_info.name.clear();
                    custom_function_info.content.clear();
                }
            }
            Ok(Event::CData(e)) => {
                if depth < 3 {
                    continue;
                }

                custom_function_info
                    .content
                    .push_str(cdata_to_string(&e).as_str());
            }
            _ => {}
        }

        buf.clear()
    }
}

fn write_custom_function_to_file(output_dir: &Path, cf: &CustomFunctionInfo) {
    let cf_filename = join_scope_id_and_name(cf.id.as_str(), cf.name.as_str());
    let cf_filename = escape_filename(&cf_filename);

    let output_file_path = output_dir.join(cf_filename).with_extension("txt");
    write_text_file(&output_file_path, &cf.content);
}
