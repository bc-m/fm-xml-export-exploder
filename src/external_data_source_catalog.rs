use std::io::{BufRead, Read};
use std::path::Path;

use quick_xml::events::{BytesStart, Event};
use quick_xml::reader::Reader;

use crate::utils::xml_utils::{
    cdata_element_to_string, end_element_to_string, start_element_to_string, text_element_to_string,
};
use crate::utils::{create_dir, write_xml_file};
use crate::Flags;

pub fn xml_extract_external_data_sources<R: Read + BufRead>(
    reader: &mut Reader<R>,
    start: &BytesStart,
    out_dir_path: &Path,
    fm_file_name: &str,
    flags: &Flags,
) {
    let out_dir_path = out_dir_path.join("external_data_sources");
    create_dir(&out_dir_path);

    let mut external_data_source_info = String::new();
    external_data_source_info.push_str(start_element_to_string(start, flags).as_str());

    let mut depth = 1;
    let mut buf = Vec::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Err(e) => {
                println!("Error {}", e);
                break;
            }
            Ok(Event::Eof) => break,
            Ok(Event::Start(e)) => {
                depth += 1;
                external_data_source_info.push_str(start_element_to_string(&e, flags).as_str());
            }
            Ok(Event::End(e)) => {
                depth -= 1;
                external_data_source_info.push_str(end_element_to_string(&e).as_str());

                if depth == 0 {
                    write_external_data_sources_to_file(
                        &out_dir_path,
                        fm_file_name,
                        &external_data_source_info,
                        flags,
                    );
                    break;
                }
            }
            Ok(Event::CData(e)) => {
                external_data_source_info.push_str(cdata_element_to_string(&e).as_str());
            }
            Ok(Event::Text(e)) => {
                external_data_source_info.push_str(text_element_to_string(&e, true).as_str());
            }
            Ok(Event::Comment(e)) => {
                external_data_source_info.push_str(text_element_to_string(&e, false).as_str());
            }
            _ => {}
        }

        buf.clear()
    }
}

fn write_external_data_sources_to_file(
    output_dir: &Path,
    fm_file_name: &str,
    content: &str,
    flags: &Flags,
) {
    let output_file_path = output_dir.join(format!("{}.xml", fm_file_name));
    write_xml_file(&output_file_path, content, 3, flags);
}
