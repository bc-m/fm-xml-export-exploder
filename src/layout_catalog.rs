use quick_xml::escape::unescape;
use std::fs;
use std::io::{BufRead, Read};
use std::path::Path;

use quick_xml::events::{BytesStart, Event};
use quick_xml::reader::Reader;

use crate::utils::attributes::get_attributes;
use crate::utils::xml_utils::{
    cdata_element_to_string, end_element_to_string, local_name_to_string, skip_element,
    start_element_to_string, text_element_to_string,
};
use crate::utils::{initialize_out_dir, write_xml_file};
use crate::Flags;
use crate::{escape_filename, join_scope_id_and_name};

#[derive(Debug, Default)]
struct LayoutInfo {
    id: String,
    name: String,
    content: String,
    path: Vec<String>,
}

pub fn xml_explode_layout_catalog<R: Read + BufRead>(
    reader: &mut Reader<R>,
    _: &BytesStart,
    out_dir_path: &Path,
    fm_file_name: &str,
    flags: &Flags,
) {
    let out_dir_path = out_dir_path.join("layouts").join(fm_file_name);
    initialize_out_dir(&out_dir_path);

    let mut layout_info = LayoutInfo::default();

    let mut depth = 1;
    let mut buf = Vec::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Err(e) => {
                println!("Error {}", e);
                break;
            }
            Ok(Event::Eof) => break,
            Ok(Event::Start(e)) => {
                depth += 1;

                if depth == 2 {
                    layout_info.id.clear();
                    layout_info.name.clear();
                    layout_info.content.clear();
                }
                if depth == 2 && e.name().as_ref() != b"Layout" {
                    skip_element(reader, &e);
                    depth -= 1;
                    continue;
                }

                if e.name().as_ref() == b"Layout" {
                    for attr in get_attributes(&e).unwrap() {
                        match attr.0.as_str() {
                            "id" => layout_info.id = attr.1.to_string(),
                            "name" => {
                                layout_info.name = unescape(attr.1.as_str()).unwrap().to_string()
                            }
                            "isFolder" => match attr.1.as_str() {
                                "True" => {
                                    layout_info.path.push(join_scope_id_and_name(
                                        layout_info.id.as_str(),
                                        layout_info.name.as_str(),
                                    ));
                                    skip_element(reader, &e);
                                    depth -= 1;
                                    continue;
                                }
                                "Marker" => {
                                    layout_info.path.pop();
                                    skip_element(reader, &e);
                                    depth -= 1;
                                    continue;
                                }
                                _ => {}
                            },
                            "isSeparatorItem" => {
                                skip_element(reader, &e);
                                depth -= 1;
                                continue;
                            }
                            _ => {}
                        }
                    }
                }

                layout_info
                    .content
                    .push_str(start_element_to_string(&e, flags).as_str());
            }
            Ok(Event::End(e)) => {
                depth -= 1;
                if depth == 0 {
                    break;
                }

                layout_info
                    .content
                    .push_str(end_element_to_string(&e).as_str());

                if depth == 1 && local_name_to_string(e.name().as_ref()) == "Layout" {
                    write_layout_to_file(&out_dir_path, &layout_info, flags)
                }
            }
            Ok(Event::CData(e)) => {
                layout_info
                    .content
                    .push_str(cdata_element_to_string(&e).as_str());
            }
            Ok(Event::Text(e)) => {
                layout_info
                    .content
                    .push_str(text_element_to_string(&e, true).as_str());
            }
            Ok(Event::Comment(e)) => {
                layout_info
                    .content
                    .push_str(text_element_to_string(&e, false).as_str());
            }
            _ => {}
        }

        buf.clear()
    }
}

fn write_layout_to_file(dir_path: &Path, layout: &LayoutInfo, flags: &Flags) {
    let layout_filename = join_scope_id_and_name(layout.id.as_str(), layout.name.as_str());
    let layout_filename = escape_filename(&layout_filename);

    // Determine output directory based on element path
    let element_path = layout
        .path
        .iter()
        .map(|e| escape_filename(e))
        .collect::<Vec<_>>()
        .join("/");

    let output_dir = dir_path.join(element_path);
    fs::create_dir_all(&output_dir)
        .unwrap_or_else(|err| panic!("Error creating directory {}: {}", output_dir.display(), err));

    let output_file_path = output_dir.join(format!("{}.xml", layout_filename));
    write_xml_file(&output_file_path, &layout.content, 4, flags);
}
