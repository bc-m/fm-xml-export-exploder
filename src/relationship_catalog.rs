use std::io::{BufRead, Read};
use std::path::Path;

use quick_xml::events::{BytesStart, Event};
use quick_xml::reader::Reader;

use crate::utils::attributes::get_attribute;
use crate::utils::xml_utils::{
    cdata_element_to_string, end_element_to_string, local_name_to_string, skip_element,
    start_element_to_string, text_element_to_string,
};
use crate::utils::{initialize_out_dir, write_xml_file};
use crate::Flags;
use crate::{escape_filename, join_scope_id_and_name};

#[derive(Debug, Default)]
struct RelationshipInfo {
    id: String,
    left: String,
    right: String,
    content: String,
}

pub fn xml_explode_relationship_catalog<R: Read + BufRead>(
    reader: &mut Reader<R>,
    _: &BytesStart,
    out_dir_path: &Path,
    fm_file_name: &str,
    flags: &Flags,
) {
    let out_dir_path = out_dir_path.join("relationships").join(fm_file_name);
    initialize_out_dir(&out_dir_path);

    let mut relationship_info = RelationshipInfo::default();
    let mut in_left = false;
    let mut in_right = false;

    let mut depth = 1;
    let mut buf = Vec::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Err(e) => {
                println!("Error {}", e);
                break;
            }
            Ok(Event::Eof) => break,
            Ok(Event::Start(e)) => {
                depth += 1;

                if depth == 2 {
                    if e.name().as_ref() == b"UUID" {
                        skip_element(reader, &e);
                        depth -= 1;
                        continue;
                    }

                    relationship_info.id.clear();
                    relationship_info.left.clear();
                    relationship_info.right.clear();
                    relationship_info.content.clear();

                    if e.name().as_ref() == b"Relationship" {
                        relationship_info.id = get_attribute(&e, "id").unwrap();
                    }
                } else if depth == 3 {
                    match e.name().as_ref() {
                        b"LeftTable" => {
                            in_left = true;
                            relationship_info.left = get_attribute(&e, "name").unwrap_or_default();
                        }
                        b"RightTable" => {
                            in_right = true;
                            relationship_info.right = get_attribute(&e, "name").unwrap_or_default();
                        }
                        _ => {}
                    }
                } else if depth == 4 && e.name().as_ref() == b"TableOccurrenceReference" {
                    if in_left {
                        relationship_info.left = get_attribute(&e, "name").unwrap();
                    } else if in_right {
                        relationship_info.right = get_attribute(&e, "name").unwrap();
                    }
                }

                relationship_info
                    .content
                    .push_str(start_element_to_string(&e, flags).as_str());
            }
            Ok(Event::End(e)) => {
                depth -= 1;
                if depth == 0 {
                    break;
                }

                relationship_info
                    .content
                    .push_str(end_element_to_string(&e).as_str());

                if depth == 1 && local_name_to_string(e.name().as_ref()) == "Relationship" {
                    write_relationship_to_file(&out_dir_path, &relationship_info, flags);
                    relationship_info.id.clear();
                    relationship_info.left.clear();
                    relationship_info.right.clear();
                    relationship_info.content.clear();
                } else if depth == 2 {
                    match e.name().as_ref() {
                        b"LeftTable" => {
                            in_left = false;
                        }
                        b"RightTable" => {
                            in_right = false;
                        }
                        _ => {}
                    }
                }
            }
            Ok(Event::CData(e)) => {
                relationship_info
                    .content
                    .push_str(cdata_element_to_string(&e).as_str());
            }
            Ok(Event::Text(e)) | Ok(Event::Comment(e)) => {
                if depth < 2 {
                    continue;
                }

                relationship_info
                    .content
                    .push_str(text_element_to_string(&e, true).as_str());
            }
            _ => {}
        }

        buf.clear()
    }
}

fn write_relationship_to_file(output_dir: &Path, relationship: &RelationshipInfo, flags: &Flags) {
    let relationship_filename = join_scope_id_and_name(
        relationship.id.as_str(),
        format!("[{}] - [{}]", relationship.left, relationship.right).as_str(),
    );
    let relationship_filename = escape_filename(&relationship_filename).replace('.', "_");
    let output_file_path = output_dir.join(format!("{}.xml", relationship_filename));
    write_xml_file(&output_file_path, &relationship.content, 4, flags);
}
