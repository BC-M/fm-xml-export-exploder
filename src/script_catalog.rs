use crate::join_scope_id_and_name;
use quick_xml::events::{BytesStart, Event};
use quick_xml::reader::Reader;
use std::collections::HashMap;
use std::io::{BufRead, Read};

use crate::utils::attributes::get_attributes;

pub fn parse_script_directories<R: Read + BufRead>(
    reader: &mut Reader<R>,
    _: &BytesStart,
) -> HashMap<String, Vec<String>> {
    let mut script_path: Vec<String> = Vec::new();
    let mut script_id_path_map: HashMap<String, Vec<String>> = HashMap::new();

    let mut depth = 1;
    let mut buf = Vec::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Err(e) => {
                println!("Error {}", e);
                break;
            }
            Ok(Event::Eof) => break,
            Ok(Event::Start(e)) => {
                depth += 1;

                if depth == 2 && e.name().as_ref() == b"Script" {
                    let mut id = String::new();
                    let mut name = String::new();
                    let mut is_folder = false;
                    let mut is_marker = false;
                    let mut is_separator = false;

                    for attr in get_attributes(&e).unwrap() {
                        match attr.0.as_str() {
                            "id" => id = attr.1.to_string(),
                            "name" => name = attr.1.to_string(),
                            "isFolder" => match attr.1.as_str() {
                                "True" => {
                                    is_folder = true;
                                }
                                "Marker" => {
                                    is_marker = true;
                                }
                                _ => {}
                            },
                            "isSeparatorItem" => is_separator = true,
                            _ => {}
                        }
                    }

                    if is_folder && !name.is_empty() {
                        script_path.push(join_scope_id_and_name(&id, &name));
                    } else if !is_folder && !is_marker && !is_separator {
                        script_id_path_map.insert(id.to_string(), script_path.clone());
                    } else if is_marker {
                        script_path.pop();
                    }
                }
            }
            Ok(Event::End(_)) => {
                depth -= 1;
                if depth == 0 {
                    break;
                }
            }
            _ => {}
        }

        buf.clear()
    }

    script_id_path_map
}
