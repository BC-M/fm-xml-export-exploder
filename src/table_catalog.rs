use crate::{escape_filename, join_scope_id_and_name};
use quick_xml::events::{BytesStart, Event};
use quick_xml::reader::Reader;
use std::io::{BufRead, Read};
use std::path::Path;

use crate::utils::attributes::get_attributes;
use crate::utils::xml_utils::{
    cdata_element_to_string, end_element_to_string, local_name_to_string, start_element_to_string,
    text_element_to_string,
};
use crate::utils::{initialize_out_dir, write_xml_file};

#[derive(Debug, Default)]
struct TableInfo {
    id: String,
    name: String,
    content: String,
}

pub fn xml_explode_table_catalog<R: Read + BufRead>(
    reader: &mut Reader<R>,
    _: &BytesStart,
    out_dir_path: &Path,
    fm_file_name: &str,
) {
    let out_dir_path = out_dir_path.join("tables").join(fm_file_name);
    initialize_out_dir(&out_dir_path);

    let mut table_info = TableInfo::default();

    let mut depth = 1;
    let mut buf = Vec::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Err(e) => {
                println!("Error {}", e);
                break;
            }
            Ok(Event::Eof) => break,
            Ok(Event::Start(e)) => {
                depth += 1;

                if depth == 2 {
                    table_info.id.clear();
                    table_info.name.clear();
                    table_info.content.clear();
                }

                if depth == 3 && e.name().as_ref() == b"BaseTableReference" {
                    for attr in get_attributes(&e).unwrap() {
                        match attr.0.as_str() {
                            "id" => table_info.id = attr.1.to_string(),
                            "name" => table_info.name = attr.1.to_string(),
                            _ => {}
                        }
                    }
                }

                table_info
                    .content
                    .push_str(start_element_to_string(&e).as_str());
            }
            Ok(Event::End(e)) => {
                depth -= 1;
                if depth == 0 {
                    break;
                }

                table_info
                    .content
                    .push_str(end_element_to_string(&e).as_str());

                if depth == 1 && local_name_to_string(e.name().as_ref()) == "FieldCatalog" {
                    write_table_to_file(&out_dir_path, &table_info);
                    table_info.id.clear();
                    table_info.name.clear();
                    table_info.content.clear();
                }
            }
            Ok(Event::CData(e)) => {
                table_info
                    .content
                    .push_str(cdata_element_to_string(&e).as_str());
            }
            Ok(Event::Text(e)) => {
                table_info
                    .content
                    .push_str(text_element_to_string(&e, true).as_str());
            }
            Ok(Event::Comment(e)) => {
                table_info
                    .content
                    .push_str(text_element_to_string(&e, false).as_str());
            }
            _ => {}
        }

        buf.clear()
    }
}

fn write_table_to_file(output_dir: &Path, table: &TableInfo) {
    let table_filename = join_scope_id_and_name(table.id.as_str(), table.name.as_str());
    let table_filename = escape_filename(&table_filename);
    let output_file_path = output_dir.join(table_filename).with_extension("xml");
    write_xml_file(&output_file_path, &table.content, 4);
}
