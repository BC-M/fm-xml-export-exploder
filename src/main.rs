use quick_xml::events::Event;
use quick_xml::reader::Reader;

use crate::base_table_catalog::parse_base_table_catalog;
use crate::custom_function_catalog::xml_explode_custom_function_catalog;
use crate::external_data_source_catalog::xml_extract_external_data_sources;
use crate::layout_catalog::xml_explode_layout_catalog;
use crate::relationship_catalog::xml_explode_relationship_catalog;
use crate::script_catalog::parse_script_directories;
use crate::script_steps_catalog::xml_explode_script_catalog;
use crate::table_catalog::xml_explode_table_catalog;
use crate::theme_catalog::xml_explode_theme_catalog;
use crate::utils::attributes::get_attribute;
use crate::value_list_catalog::xml_explode_value_list_catalog;
use encoding_rs_io::DecodeReaderBytes;
use rayon::prelude::*;
use std::collections::HashMap;
use std::{
    env,
    error::Error,
    fs,
    fs::File,
    io::BufReader,
    path::{Path, PathBuf},
    time::Instant,
};

mod base_table_catalog;
mod calculations;
mod custom_function_catalog;
mod external_data_source_catalog;
mod layout_catalog;
mod relationship_catalog;
mod script_catalog;
mod script_steps;
mod script_steps_catalog;
mod table_catalog;
mod theme_catalog;
mod utils;
mod value_list_catalog;

fn main() -> Result<(), Box<dyn Error>> {
    let start = Instant::now();

    let in_dir = env::args().nth(1).ok_or("Input directory not provided")?;
    let out_dir = env::args().nth(2).ok_or("Output directory not provided")?;

    let in_dir = Path::new(&in_dir);
    if !is_valid_directory(in_dir) {
        println!("'{}' is not a valid directory.", in_dir.display());
        return Ok(());
    }

    let out_dir = Path::new(&out_dir);
    if !is_valid_directory(out_dir) {
        println!("'{}' is not a valid directory.", out_dir.display());
        return Ok(());
    }

    // Read directory contents
    let paths = fs::read_dir(in_dir)?
        .filter_map(|entry| entry.ok().map(|e| e.path())) // Filter out directories and unwrap results
        .filter(|path| path.is_file() && path.extension().unwrap_or_default() == "xml") // Filter XML files
        .collect::<Vec<_>>(); // Collect paths into a vector

    println!("Start processing {} files...", paths.len());

    // Process XML files in parallel
    paths.par_iter().for_each(|path| explode_xml(path, out_dir));

    let duration = start.elapsed();
    if duration.as_secs() > 9 {
        println!("Completed in {:?} seconds.", duration.as_secs());
    } else {
        println!("Completed in {:?} ms.", duration.as_millis());
    }

    Ok(())
}

fn is_valid_directory(dir_path: &Path) -> bool {
    fs::metadata(dir_path)
        .map(|metadata| metadata.is_dir())
        .unwrap_or(false)
}

fn explode_xml(fm_export_file_path: &PathBuf, out_dir_path: &Path) {
    let start = Instant::now();
    let fm_export_file_name = fm_export_file_path.file_name().unwrap().to_str().unwrap();

    // Open XML file
    let file = match File::open(fm_export_file_path) {
        Ok(file) => file,
        Err(err) => {
            eprintln!(
                "Error opening file {}: {}",
                fm_export_file_path.display(),
                err
            );
            return;
        }
    };

    // Initialize variables
    let mut fm_file_name = String::new();
    let mut depth = 0;
    let mut script_id_path_map: HashMap<String, Vec<String>> = HashMap::new();
    let mut table_name_id_map: HashMap<String, String> = HashMap::new();

    // Iterate over XML events
    let decode_reader = BufReader::new(DecodeReaderBytes::new(file));
    let mut reader = Reader::from_reader(decode_reader);
    let mut buf: Vec<u8> = Vec::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Err(e) => {
                println!("Error in {}: {}", fm_export_file_path.display(), e);
                break;
            }
            Ok(Event::Eof) => break,
            Ok(Event::Start(e)) => {
                match depth {
                    0 => match e.name().as_ref() {
                        b"FMDynamicTemplate" | b"FMSaveAsXML" => {
                            fm_file_name = get_attribute(&e, "File")
                                .unwrap()
                                .strip_suffix(".fmp12")
                                .unwrap()
                                .to_string()
                        }
                        _ => {
                            eprintln!("Unsupported XML-format in file {}", fm_export_file_name);
                            return;
                        }
                    },
                    2 => {
                        if e.name().as_ref() == b"ModifyAction" {
                            break;
                        }
                    }
                    3 => match e.name().as_ref() {
                        b"BaseTableCatalog" => {
                            table_name_id_map = parse_base_table_catalog(&mut reader, &e);
                            continue;
                        }
                        b"LayoutCatalog" => {
                            xml_explode_layout_catalog(
                                &mut reader,
                                &e,
                                out_dir_path,
                                &fm_file_name,
                            );
                            continue;
                        }
                        b"FieldsForTables" => {
                            xml_explode_table_catalog(
                                &mut reader,
                                &e,
                                out_dir_path,
                                &fm_file_name,
                                &table_name_id_map,
                            );
                            continue;
                        }
                        b"CalcsForCustomFunctions" => {
                            xml_explode_custom_function_catalog(
                                &mut reader,
                                &e,
                                out_dir_path,
                                &fm_file_name,
                            );
                            continue;
                        }
                        b"StepsForScripts" => {
                            xml_explode_script_catalog(
                                &mut reader,
                                &e,
                                out_dir_path,
                                &fm_file_name,
                                &script_id_path_map,
                            );
                            continue;
                        }
                        b"ScriptCatalog" => {
                            script_id_path_map = parse_script_directories(&mut reader, &e);
                            continue;
                        }
                        b"ExternalDataSourceCatalog" => {
                            xml_extract_external_data_sources(
                                &mut reader,
                                &e,
                                out_dir_path,
                                &fm_file_name,
                            );
                            continue;
                        }
                        b"ValueListCatalog" => {
                            xml_explode_value_list_catalog(
                                &mut reader,
                                &e,
                                out_dir_path,
                                &fm_file_name,
                            );
                            continue;
                        }
                        b"RelationshipCatalog" => {
                            xml_explode_relationship_catalog(
                                &mut reader,
                                &e,
                                out_dir_path,
                                &fm_file_name,
                            );
                            continue;
                        }
                        b"ThemeCatalog" => {
                            xml_explode_theme_catalog(&mut reader, &e, out_dir_path, &fm_file_name);
                            continue;
                        }
                        _ => {}
                    },
                    _ => {}
                }
                depth += 1;
            }
            Ok(Event::End(_)) => {
                depth -= 1;
            }
            _ => {}
        }
        buf.clear()
    }

    println!(
        "{} finished in {} ms.",
        fm_export_file_name,
        start.elapsed().as_millis()
    );
}

fn join_scope_id_and_name(scope_id: &str, scope_name: &str) -> String {
    format!("{} - ID {}", scope_name, scope_id)
}

fn escape_filename(filename: &str) -> String {
    // Replace special characters with underscores
    let escaped_filename = filename
        .chars()
        .map(|c| match c {
            '/' | '\\' | ':' | '*' | '?' | '"' | '<' | '>' | '|' | '{' => '_',
            _ => c,
        })
        .collect::<String>();

    escaped_filename.trim().to_string()
}

fn should_skip_line(line: &str) -> bool {
    // de-noise TODO: make optional
    if line.contains("<TagList></TagList>") {
        return true;
    };
    if line.contains("<OwnerID></OwnerID>") {
        return true;
    };
    if line.contains("<Options>0</Options>") {
        return true;
    }; // Default
    if line.contains("<Options>1048576</Options>") {
        return true;
    }; // Breakpoint
    if line.contains("<UUID>") && line.contains("</UUID>") {
        return true;
    };

    false
}

#[cfg(test)]
mod tests {
    use super::*;
    use fs::remove_dir_all;
    use insta::assert_snapshot;
    use similar_asserts::assert_eq;
    use std::io::Read;
    use walkdir::WalkDir;

    #[test]
    fn test_escape_filename() {
        assert_eq!(escape_filename("filename.xml"), "filename.xml");
        assert_eq!(escape_filename("file/name.xml"), "file_name.xml");
        assert_eq!(escape_filename("file|name.xml"), "file_name.xml");
    }

    #[test]
    fn snapshot_test() {
        let snapshot_dir = Path::new("./tests/snapshots");
        let input_dir = Path::new("./tests/xml");
        let output_dir = Path::new("./out");
        let _ = remove_dir_all(output_dir);

        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_path(snapshot_dir);
        settings.set_prepend_module_to_snapshot(false);

        let paths = fs::read_dir(input_dir)
            .unwrap()
            .filter_map(|entry| entry.ok().map(|e| e.path()))
            .filter(|path| path.is_file())
            .filter(|e| e.file_name().unwrap() != ".DS_Store")
            .collect::<Vec<_>>();

        for path in paths {
            explode_xml(&path, output_dir);
        }

        let output_files: Vec<PathBuf> = WalkDir::new(output_dir)
            .into_iter()
            .filter_map(|entry| entry.ok())
            .filter(|e| e.path().is_file())
            .filter(|e| e.file_name().to_str().unwrap() != ".DS_Store")
            .map(|e| e.path().to_path_buf())
            .collect();

        let mut output_file_paths = output_files
            .iter()
            .map(|file| file.strip_prefix(output_dir).unwrap())
            .map(|file| file.to_string_lossy())
            .collect::<Vec<_>>();
        output_file_paths.sort();

        let snapshot_files: Vec<PathBuf> = WalkDir::new(snapshot_dir)
            .into_iter()
            .filter_map(|entry| entry.ok())
            .filter(|e| e.path().is_file())
            .filter(|e| e.file_name().to_str().unwrap().ends_with(".snap"))
            .map(|e| e.path().to_path_buf())
            .collect();

        let mut snapshot_file_paths = snapshot_files
            .iter()
            .map(|file| file.strip_prefix(snapshot_dir).unwrap())
            .map(|file| {
                file.to_string_lossy()
                    .strip_suffix(".snap")
                    .unwrap()
                    .to_string()
            })
            .collect::<Vec<_>>();
        snapshot_file_paths.sort();

        assert_eq!(output_file_paths.join("\n"), snapshot_file_paths.join("\n"));

        for output_file in &output_files {
            let output_content = String::from_utf8(read_file(output_file)).unwrap();
            let output_file_slug = output_file.strip_prefix(output_dir).unwrap();

            settings.set_snapshot_path(
                Path::new("../tests/snapshots")
                    .join(output_file_slug)
                    .parent()
                    .unwrap(),
            );

            settings.bind(|| {
                assert_snapshot!(output_file.file_name().unwrap().to_str(), output_content);
            });
        }
    }

    fn read_file(file_path: &PathBuf) -> Vec<u8> {
        let mut file = fs::File::open(file_path).expect("Failed to open file");
        let mut content = Vec::new();
        file.read_to_end(&mut content).expect("Failed to read file");
        content
    }
}
