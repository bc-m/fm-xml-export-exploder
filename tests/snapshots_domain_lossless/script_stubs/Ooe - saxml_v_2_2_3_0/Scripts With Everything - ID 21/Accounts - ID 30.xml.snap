---
source: src/main.rs
expression: output_content
---
<Script id="30" name="Accounts">
	<UUID modifications="2" userName="Mislav" accountName="admin" timestamp="2025-06-20T18:24:25">5D8583E9-201F-4B93-9A84-4991B1A8A837</UUID>
	<Options hidden="False" access="ReadWrite" SiriShortcutVisible="False" runwithfullaccess="False" compatibility="0">0</Options>
	<TagList></TagList>
</Script>
